browser/components/newtab/NewTabPrefsProvider.jsm

```diff
@@ -31,7 +31,24 @@ PrefsProvider.prototype = {
   observe(subject, topic, data) { // jshint ignore:line
     if (topic === "nsPref:changed") {
       if (prefsSet.has(data)) {
-        this.emit(data);
+        switch (data) {
+          case "browser.newtabpage.enabled":
+            this.emit("setEnabled", {
+              enabled: Services.prefs.getBoolPref(data),
+            });
+            break;
+          case "browser.newtabpage.enhanced":
```

Comment:
I'm fine either way. One way to see it is... one approach has less code than the other ;-)